content/articles/ansible-and-nix.md

```diff
@@ -0,0 +1,333 @@
+
+If you want to know more about Ansible, I recommend starting with the official
+[How Ansible Works](http://www.ansible.com/how-ansible-works) article.
+
+The reality of setting up a system, even with a configuration management tool,
+is usually a trial and error process. The following could be an example:
+
+* Install database package
+* Copy database configuration file
+* Test database
+* Fix anything wrong in the configuration and re-deploy
+* Refactor the configuration
+* Test it again
+* Move on to the configuration of the web server
+
+Using a tool like Ansible, this process exhibits a major flaw. If you were to
+remove the line to install a package in an Ansible playbook, the package would
+still be installed on the server you are configuring. Which means that after a
+refactoring phase the server will work properly but with an incomplete
+configuration. This problem is not limited to packages, but also to the
+presence of users, files, their permissions, their content... Anything you
+could configure with Ansible.
+
+The result of this flaw is that even if you managed to get a server properly
+configured once, you don't know if it will work again.
+If you change your server configuration over time without creating new
+servers, then you need to make sure your configuration files does not diverge
+from the actual state of the server.
+
+Also, even if you tested your playbooks well, it is still possible for them to
+stop working if a system package or application dependencies has been updated
+in your system.
+
+## Nix
+
+This is where the Nix project comes in. It consists of [Nix, The Purely
+Functional Package Manager](https://nixos.org/nix/), [NixOS, The Purely
+Functional Linux Distribution](https://nixos.org/) and [NixOps, The NixOS Cloud
+Deployment Tool](https://nixos.org/nixops/).
+
+Nix produces packages that link to all their dependencies at specific versions
+and that are entirely independent from each other. The result from these
+properties means that you cannot get into the problem illustrated previously. 
+
+NixOS is a simple and modern Linux distribution configured declaratively using
+the same language as Nix packages.
+
+Finally, NixOps is used to configure a «network» of NixOS systems, still using
+the same language.
+
+The learning curve for those tools is steep, the official documentation is
+light and there aren't that many resources available on the web about Nix so I
+had to learn it mostly by reading the source of the tools. Also, NixOps being a
+young project I hit a few open bugs.
+
+The [Nix manual](https://nixos.org/nix/manual/) is a good start to learn about
+the package manager. This [Nix by example
+article](https://medium.com/@MrJamesFisher/nix-by-example-a0063a1a4c55)
+describes the Nix expression language in more detail.
+To understand the rationale behind Nix and NixOS, you can read the accessible
+[NixOS: A Purely Functional Linux
+Distribution](http://nixos.org/~eelco/pubs/nixos-jfp-final.pdf) paper.
```

Comment:
i would put the paper in the link as well, otherwise that's the same link as the above to https://nixos.org/